source/server/lds_api.cc

```diff
@@ -64,8 +68,10 @@ void LdsApiImpl::onConfigUpdate(const ResourceVector& resources, const std::stri
         ENVOY_LOG(debug, "lds: add/update listener '{}' skipped", listener_name);
       }
     } catch (const EnvoyException& e) {
-      throw EnvoyException(
-          fmt::format("Error adding/updating listener {}: {}", listener_name, e.what()));
+      // retry after the remove opertions are performed to allow new listeners
+      // with addresses used by other listeners that have to be removed.
+      listeners_to_retry.push_back(listener);
```

Comment:
I'm not in favor of this generic retry solution, since it might retry for genuine failures which we should not do.

Why can't we just do the removals first?

Reply:
@mattklein123 Removals first can certainly be done. The reason I did not do it is that the existing code goes a bit out of its way to not do that. As I mentioned on the bug issue, my guess is that that was an optimization done to avoid removing listeners unnecessarily in more common cases. Is it the most common case that there are a lot of listeners that remain after an update and removing them first and adding them again creates too much of an overhead? I don't know the answer to that.

The addOrUpdateListener() throws exception in a couple of places. One possibility is having a specific exception when the reason for failure is a duplicate address and do the retry only for that specific throw. But even this may result in unnecessary retries.

Either way, let me know which sounds more reasonable and I'll adjust the patch.

Reply:
Just precompute which listeners need to be removed, remove them, then do the adds? There should be no unnecessary removals.

Reply:
Yes, that makes sense.